File: backend/app/laptime.py
"""Pure lap-time parsing: `m:ss.ms` or `ss.ms` strings to integer milliseconds."""

import re
from typing import Final

__all__ = ["LapTimeParseError", "parse_lap_time_ms"]


class LapTimeParseError(ValueError):
    """Raised when a lap-time string cannot be parsed into milliseconds."""

    raw: str

    def __init__(self, raw: str) -> None:
        self.raw = raw
        super().__init__(f"invalid lap time: {raw!r}")


_LAP_TIME_PATTERN: Final[re.Pattern[str]] = re.compile(
    r"^(?:(\d+):([0-5]?\d)|(\d{1,3}))(?:\.(\d{1,3}))?$"
)


def parse_lap_time_ms(raw: str) -> int:
    """Parse a lap-time string (`m:ss.ms` or `ss.ms`) into integer milliseconds.

    The fraction may use 1-3 digits and is right-padded to milliseconds
    (`.1` = 100 ms, `.50` = 500 ms, `.132` = 132 ms). Surrounding whitespace
    is tolerated. Raises LapTimeParseError for anything unparseable.
    """
    match = _LAP_TIME_PATTERN.fullmatch(raw.strip())
    if match is None:
        raise LapTimeParseError(raw)
    minutes_str, seconds_str, bare_seconds_str, fraction_str = match.groups()
    minutes = int(minutes_str) if minutes_str is not None else 0
    seconds = int(seconds_str or bare_seconds_str or "0")
    fraction_ms = int(fraction_str.ljust(3, "0")) if fraction_str is not None else 0
    return (minutes * 60 + seconds) * 1000 + fraction_ms

File: frontend/src/lib/laptime.ts
/** Lap-time parsing/formatting, mirroring `backend/app/laptime.py` exactly. */

const LAP_TIME_PATTERN = /^(?:(\d+):([0-5]?\d)|(\d{1,3}))(?:\.(\d{1,3}))?$/;

export type LapTimeParseResult =
  | { readonly ok: true; readonly ms: number }
  | { readonly ok: false; readonly message: string };

/**
 * Parse a lap-time string (`m:ss.ms` or `ss.ms`) into integer milliseconds.
 * The fraction may use 1-3 digits and is right-padded to milliseconds
 * (`.1` = 100 ms, `.50` = 500 ms, `.132` = 132 ms). Surrounding whitespace
 * is tolerated. Mirrors the backend regex and padding rules.
 */
export function parseLapTime(raw: string): LapTimeParseResult {
  const match = LAP_TIME_PATTERN.exec(raw.trim());
  if (match === null) {
    return {
      ok: false,
      message: "must be m:ss.ms or ss.ms — e.g. 1:03.500 or 40.132",
    };
  }
  const [, minutesRaw, secondsWithColonRaw, secondsBareRaw, fractionRaw] = match;
  const minutes = minutesRaw === undefined ? 0 : Number(minutesRaw);
  const seconds = Number(secondsWithColonRaw ?? secondsBareRaw ?? "0");
  const fractionMs =
    fractionRaw === undefined ? 0 : Number(fractionRaw.padEnd(3, "0"));
  return { ok: true, ms: (minutes * 60 + seconds) * 1000 + fractionMs };
}

/** Format integer milliseconds as `m:ss.ms` (or `ss.ms` under a minute). */
export function formatLapTime(ms: number): string {
  const totalSeconds = Math.floor(ms / 1000);
  const fraction = ms % 1000;
  const minutes = Math.floor(totalSeconds / 60);
  const seconds = totalSeconds % 60;
  const fractionText = String(fraction).padStart(3, "0");
  if (minutes > 0) {
    return `${minutes}:${String(seconds).padStart(2, "0")}.${fractionText}`;
  }
  return `${seconds}.${fractionText}`;
}
